pkg/service/healthcheck/service_integration_test.go

```diff
@@ -547,6 +553,36 @@ func tryStartAgent(t *testing.T, hosts []string) {
 	}
 }
 
+func ensureNodesAreUP(t *testing.T, hosts []string, timeout time.Duration) error {
+	t.Helper()
+
+	done := make(chan struct{})
+	wg := sync.WaitGroup{}
+	for _, host := range hosts {
+		wg.Add(1)
+		go func(h string) {
+			defer wg.Done()
+
+			err := WaitForNodeUP(h, timeout)
+			if err != nil {
+				t.Log(err)
+			}
+		}(host)
+	}
+
+	go func() {
+		wg.Wait()
+		close(done)
+	}()
+
+	select {
+	case <-done:
+		return nil
```

Comment:
So the error (that could come from the fact that a node is down) is ignored (except for being logged), so the whole function can return nil even when nodes are down?

Reply:
Yes, error is just logged. It comes from the fact that WaitForNodeUP may only return error when it times out.
But the logic is blurry, I agree. Even though that the function will not return that node is UP when it's DOWN, but it will return timeout error.

Anyway, let me try to make the logic bit more easy to read and understand.

Reply:
> Even though that the function will not return that node is UP when it's DOWN, but it will return timeout error.

But the timeout error in this case is synonymous with node not being in UN state, right?
And then it's ignored - which means that this function has a race between the timeout set within its body and the timeout set in the per-node check. So it could return no error when all nodes are down and timeout on UN check.